Introduce ShardRouting.Role

DaveCTurner opened the pull request:
Introduces the concept of a `ShardRouting.Role` which allows plugins to assign
different roles to each shard copy in a replication group. Unlike the
primary/replica flag, a shard copy role sticks to the shard copy throughout its
life and is preserved across relocations and reassignments.

DaveCTurner commented:
## Reviewer guidance

This is a pretty noisy change due to the volume of plumbing and test changes needd. Here's some notes to help navigate it:

1. We introduce a `ShardCopyRole` which defines the role of each `ShardRouting`. Roles are stickier than things like the primary/replica flag. At the moment a role defines whether a shard copy is promotable to primary, and whether it's exposed to searches, but once the concept is in place we might want to refine this a little. By default all shard copies have an "empty" role.

2. We introduce a `ShardCopyRoleFactory` which allows plugins to define the `ShardCopyRole` when creating a `ShardRouting`. At the moment we distinguish the three different ways that a `ShardRouting` is created, with a little bit of context, but again once the concept is in place there will be scope to refine this as needed.

3. We (slightly ab)use the `NamedWriteable` framework to allow plugins to control the wire serialization of `ShardRouting`. By default, the empty role takes no bytes on the wire.

4. We adjust `RoutingNodes` and `IndexShardRoutingTable` to take the promotability flag into account when selecting a shard to promote to primary, and also when choosing a replica to remove.

5. We adjust `SearchShardIterator` to take the searchability flag into account when selecting shards to search.

6. There's some new assertions in `ShardRouting`, `IndexShardRoutingTable`, and `RoutingNodes` to verify new invariants, but there's currently only one test that only exercises this functionality, `ShardCopyRoleIT`.

DaveCTurner commented:
In terms of possible alternative approaches, as Henning suggested we could avoid making `ShardCopyRole` pluggable and instead add properties as needed to `ShardRouting` directly. I don't think that simplifies things much, because we'd still have to do most of this work to allow the plugin to control the behaviour as each `ShardRouting` is created. It would also reduce our freedom to change the properties over time because of the BwC/wire compatibility concerns.

In practice we want to use this for a binary thing (in today's plan at least) so we could also just add a single opaque flag to `ShardRouting`, and leave it up to the plugin to determine what the flag means. Again I don't think this simplifies things much, and I worry that this approach makes the codebase more obscure, harder to work with, and thus risks introducing more bugs.

idegtiarenko commented on server/src/main/java/org/elasticsearch/cluster/ClusterModule.java:
```diff
@@ -427,4 +475,52 @@ public void setExistingShardsAllocators(GatewayAllocator gatewayAllocator) {
+        return List.of();
+    }
+
+    private static class EmptyShardCopyRole implements ShardCopyRole {
```
I wonder if we should name it default to emphasize no special behavior?

DaveCTurner commented:
There's various tests which will fail the new invariants it seems; I extracted a PR just to add the invariants and fix up all the tests to reduce noise here: https://github.com/elastic/elasticsearch/pull/92705

DaveCTurner committed:
Introduce roles for `ShardRouting` instances

Introduces the concept of a `ShardRouting.Role` which allows plugins to
assign different roles to each shard copy in a replication group. Unlike
the primary/replica flag, a shard copy role sticks to the shard copy
throughout its life and is preserved across relocations and
reassignments.

DaveCTurner committed:
Adopt ShardRouting.Role

DaveCTurner committed:
Fix compile

henningandersen reviewed (commented):
This direction looks great. With sufficient testing, this looks ready for me.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java:
```diff
@@ -655,17 +654,36 @@ assert node(failedShard.currentNodeId()).getByShardId(failedShard.shardId()) ==
+            for (ShardRouting unpromotableReplica : List.copyOf(assignedShards(failedPrimary.shardId()))) {
+                assert unpromotableReplica.primary() == false : unpromotableReplica;
+                assert unpromotableReplica.isPromotableToPrimary() == false : unpromotableReplica;
+                moveToUnassigned(
```
I think we do rely on this invariant today in various places, although it's been an implicit assumption for such a long time that it's hard to know exactly where those places might be. It is verified by `IndexShardRoutingTable.Builder#noDuplicatePrimary` which was added pretty recently, but this assertion only didn't hold in a few rather specific unit tests. Mainly I think that's because of the `ReplicaAfterPrimaryActiveAllocationDecider` which would prevent the allocation of primary-free replicas in all realistic scenarios.

I'm pretty sure we can adjust things to cope with this requirement anyway, but it'll be a bit of an adventure and one best left for another day.

DaveCTurner committed:
Merge branch 'main' into 2023-01-04-shard-copy-role

idegtiarenko commented on server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -855,4 +894,66 @@ public long getExpectedShardSize() {
+
+    public Role getRole() {
+        return role;
+    }
```
Most of the accessors are named without `get` prefix in this class. Consider renaming to `role()`

DaveCTurner replied:
:+1: see [36e758f](https://github.com/elastic/elasticsearch/pull/92668/commits/36e758fcf96229d8a0996f5ed65c5eea2404add6)

DaveCTurner committed:
Add integ tests

DaveCTurner committed:
Rename accessor

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

DaveCTurner committed:
Merge branch 'main' into 2023-01-04-shard-copy-role

DaveCTurner committed:
Remove watcher at end of each test

DaveCTurner committed:
Update docs/changelog/92668.yaml

elasticsearchmachine commented:
Hi @DaveCTurner, I've created a changelog YAML for you.

DaveCTurner committed:
Merge branch 'main' into 2023-01-04-shard-copy-role

DaveCTurner committed:
TODOs done/tracked elsewhere

DaveCTurner committed:
Comments/reorderings

idegtiarenko commented on docs/changelog/92668.yaml:
```diff
@@ -0,0 +1,5 @@
+pr: 92668
+summary: "[WIP] Shard copy role experiments"
```
```suggestion
summary: "Introduce ShardRouting.Role"
```

idegtiarenko commented on server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -770,6 +803,7 @@ public int hashCode() {
             h = 31 * h + (allocationId != null ? allocationId.hashCode() : 0);
             h = 31 * h + (unassignedInfo != null ? unassignedInfo.hashCode() : 0);
             h = 31 * h + (relocationFailureInfo != null ? relocationFailureInfo.hashCode() : 0);
+            h = 31 * h + role.hashCode();
```
NIT: should we replace all above with `Objects.hash(..)`?

DaveCTurner commented:
@elasticmachine please run elasticsearch-ci/bwc

idegtiarenko reviewed (approved):
Left some comments, but otherwise :+1:

DaveCTurner committed:
Update docs/changelog/92668.yaml

Co-authored-by: Ievgen Degtiarenko <ievgen.degtiarenko@gmail.com>

DaveCTurner committed:
Javadocs

DaveCTurner committed:
Merge branch 'main' into 2023-01-04-shard-copy-role

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/ClusterModule.java:
```diff
@@ -138,10 +141,36 @@ public ClusterModule(
+                public ShardRouting.Role newEmptyRole(int copyIndex) {
+                    return ShardRouting.Role.DEFAULT;
+                }
+            };
```
We discussed this in another channel; I added some Javadocs explaining it in 7616fcccde486ac10c21ead9085927bdc75bb84e.
